controller/src/test/java/io/pravega/controller/rest/v1/HealthTests.java

```diff
@@ -0,0 +1,502 @@
+import io.pravega.shared.health.HealthService;
+
+import io.pravega.shared.health.HealthServiceFactory;
+import io.pravega.shared.health.Status;
+import io.pravega.shared.health.impl.HealthConfigImpl;
+import io.pravega.shared.health.impl.StatusAggregatorImpl;
+import io.pravega.test.common.TestUtils;
+import lombok.extern.slf4j.Slf4j;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.Timeout;
+
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.UriBuilder;
+import java.net.URI;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import static org.mockito.Mockito.mock;
+
+/**
+ * The {@link HealthTests} can be divided into two sets: those that test the implicit component (root) that is automatically
+ * created alongside the {@link io.pravega.shared.health.ContributorRegistry} and those that test specific (non-root) HealthComponents.
+ *
+ * Both implicit and explicit requests have the same code path(s) so we can expect results to be similar between each
+ * class of tests.
+ */
+@Slf4j
+public class HealthTests {
+
+    private static final String HOST = "localhost";
+
+    private static final String IMPLICIT_INDICATOR = "implicit-health-child";
+
+    private static final String IMPLICIT_COMPONENT = "health-component";
+
+    //Ensure each test completes within 30 seconds.
+    @Rule
+    public final Timeout globalTimeout = new Timeout(10000, TimeUnit.SECONDS);
+
+    private  HealthServiceFactory healthServiceFactory;
+
+    private RESTServerConfig serverConfig;
+    private RESTServer restServer;
+    private Client client;
+    private ConnectionFactory connectionFactory;
+    private HealthService service;
+
+    @Before
+    public void setup() throws Exception {
+        ControllerService mockControllerService = mock(ControllerService.class);
+        serverConfig = getServerConfig();
+        connectionFactory = new SocketConnectionFactoryImpl(ClientConfig.builder().build());
+        healthServiceFactory = new HealthServiceFactory(Optional.ofNullable(getHealthConfig()));
+        service = healthServiceFactory.createHealthService(true);
+        restServer = new RESTServer(null, mockControllerService, null, serverConfig,
+                connectionFactory, service);
+        restServer.startAsync();
+        restServer.awaitRunning();
+        client = createJerseyClient();
+        // Must provide a HealthIndicator to the HealthComponent(s) defined by the HealthConfig.
+        service.registry().register(new StaticHealthyIndicator(IMPLICIT_INDICATOR), IMPLICIT_COMPONENT);
+    }
+
+    @After
+    public void teardown() {
+        service.clear();
```

Comment:
shut down all services that you started in `setup`. Otherwise you leave things hanging and you'll have a leak.